File: leetcode/go/1004.max-consecutive-ones-iii.go
func longestOnes(nums []int, k int) int {
    left, ans, count := 0, 0, 0
    for right, n := range nums {
        if n == 0 {
            count++
        }
        for count > k {
            if nums[left] == 0 {
                count--
            }
            left++
        }
        ans = max(ans, right-left+1)
    }
    return ans
}

func max(a, b int) int {
    if a < b { return b }
    return a
}

File: leetcode/python/1004.max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        l, r, flips, ans = 0, 0, 0, 0
        
        while r < len(nums):
            if nums[r] == 0:
                flips += 1
            while flips > k:
                if nums[l] == 0:
                    flips -= 1
                l += 1
            ans = max(ans, r - l + 1)
            r += 1
        return ans

